### ISSUE TITLE
unknown url type: urn:

### ISSUE THREAD

**Issue by mshonichev** [sissaschool/xmlschema#377]
Hello!

I'm trying to validate OASIS DITAMAP XML file against bookmap DTD they provide with their DITA-OT package

however, schema is very big and split over multiple files, each referencing other via 'urn:...' notation

so, naive
```
>>> import xmlschema
>>> my_schema = xmlschema.XMLSchema("bookmap.xsd")
```

immediately fail with 
```
<stdin>:1: XMLSchemaIncludeWarning: Include schema failed: cannot access to resource 'urn:oasis:names:tc:dita:xsd:mapGroupMod.xsd:1.2': unknown url type: urn.
<stdin>:1: XMLSchemaIncludeWarning: Include schema failed: cannot access to resource 'urn:oasis:names:tc:dita:xsd:commonElementMod.xsd:1.2': unknown url type: urn.
<stdin>:1: XMLSchemaIncludeWarning: Include schema failed: cannot access to resource 'urn:oasis:names:tc:dita:xsd:tblDeclMod.xsd:1.2': unknown url type: urn.
...
```

I got all the files at hand locally.

Is there any chance to configure custom URL scheme 'urn' resolver to 'download' referenced dependency from the local folder?

Thank you in advance, the exact traceback was:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/xmlschema/resources.py", line 302, in fetch_resource
    with urlopen(url, timeout=timeout):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/urllib/request.py", line 519, in open
    response = self._open(req, data)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/urllib/request.py", line 541, in _open
    return self._call_chain(self.handle_open, 'unknown',
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/urllib/request.py", line 496, in _call_chain
    result = func(*args)
             ^^^^^^^^^^^
  File "/usr/local/lib/python3.11/urllib/request.py", line 1419, in unknown_open
    raise URLError('unknown url type: %s' % type)
urllib.error.URLError: <urlopen error unknown url type: urn>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/xmlschema/validators/schemas.py", line 1187, in _parse_inclusions
    schema = self.include_schema(location, self.base_url)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/xmlschema/validators/schemas.py", line 1231, in include_schema
    schema_url = fetch_resource(location, base_url)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/xmlschema/resources.py", line 309, in fetch_resource
    raise XMLResourceError(msg % {'url': url, 'reason': err.reason})
xmlschema.exceptions.XMLResourceError: cannot access to resource 'urn:oasis:names:tc:dita:xsd:mapGrp.xsd:1.2': unknown url type: urn

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.11/site-packages/xmlschema/validators/schemas.py", line 463, in __init__
    self._parse_inclusions()
  File "/usr/local/lib/python3.11/site-packages/xmlschema/validators/schemas.py", line 1195, in _parse_inclusions
    self.parse_error(err, child)
  File "/usr/local/lib/python3.11/site-packages/xmlschema/validators/xsdbase.py", line 190, in parse_error
    raise error
xmlschema.validators.exceptions.XMLSchemaParseError: cannot access to resource 'urn:oasis:names:tc:dita:xsd:mapGrp.xsd:1.2': unknown url type: urn:

Schema:

  <xs:redefine xmlns:xs="http://www.w3.org/2001/XMLSchema" schemaLocation="urn:oasis:names:tc:dita:xsd:mapGrp.xsd:1.2">
      <xs:group name="topicref">
          <xs:choice>
              <xs:group ref="topicref" />
              <xs:group ref="mapgroup-d-topicref" />
          </xs:choice>
      </xs:group>
  </xs:redefine>

Path: /xs:schema/xs:redefine[1]


```

**Comment by brunato** [comment]
Hi,
not yet, this requires an enhancement of XML resource module implementation. 

I could wrap *urlopen()* calls in a callable that resolves registred 'urn:' URLs to another effective one, but any advice for doing this is welcome.

thank you

**Comment by brunato** [comment]
I think about replacing _urlopen_ calls with calls to an helper function that checks the URL scheme and than uses a module level dict to map the URL to an effective URL (local or not). 

A register function can be defined to populate or remove entries from this global dict.

Another way, more complex to implement but more flexible, can be adding another option (a dict) to XMLSchema and XMLResource classes to support custom URL/URN resolving for each schema/resource instance. Also in this case an helper function for replacing _urlopen_ is necessary.

**Comment by mshonichev** [comment]
Good news everyone! :)

I've found OASIS DITA XSD without the 'urn:' embeddings inside the documents.

See `schema-url` in their package
http://docs.oasis-open.org/dita/dita/v1.3/errata02/os/complete/part3-all-inclusive/dita-v1.3-errata02-os-part3-all-inclusive-complete-grammars.zip

`xmlschema` parses them flawlessly!

The only catch was that due to enormous size of the model, recursion limit must be extended:

Following are complete working example of linting `.dita` and `.ditamap` files with `xmlschema` 
just bump your files into `input` directory and off you go!

```python
from pathlib import Path
from lxml import etree
from xmlschema import limits, XMLSchema

# default value 15 is not enough for DITA model!
limits.MAX_MODEL_DEPTH = 20

schema = Path(__file__).parent / "schema-url"
bookmap_schema_path = schema / "bookmap" / "xsd" / "bookmap.xsd"
topic_schema_path = schema / "technicalContent" / "xsd" / "topic.xsd"

# path to directory with input files to lint
@pytest.fixture
def input_path() -> Path:
    return Path(__file__).parent / "input"

# names of .ditamap files to lint
@pytest.fixture(scope="function", params=[
    'index.ditamap',
])
def ditamap_name(request: pytest.FixtureRequest) -> str:
    return getattr(request, "param", 'index.ditamap')

# names of .dita files to lint
@pytest.fixture(scope="function", params=[
    'about.dita',
    'index.dita',
    'manual.dita',
])
def dita_name(request: pytest.FixtureRequest) -> str:
    return getattr(request, "param", 'index.dita')


@pytest.fixture(scope="package")
def bookmap_schema() -> xmlschema.XMLSchema:
    return XMLSchema(bookmap_schema_path)


@pytest.fixture(scope="package")
def topic_schema() -> xmlschema.XMLSchema:
    return XMLSchema(topic_schema_path)


@pytest.fixture(scope="function")
def ditamap(input_path: Path, ditamap_name: str) -> etree.ElementTree:
    return cast(etree.ElementTree, etree.parse(str(input_path / ditamap_name)))


@pytest.fixture(scope="function")
def dita(input_path: Path, dita_name: str) ->etree.ElementTree:
    return cast(etree.ElementTree, etree.parse(str(input_path / dita_name)))


def test_bookmap_schema(input_path: Path, ditamap: etree.ElementTree, bookmap_schema: XMLSchema) -> None:
    bookmap_schema.validate(ditamap)


def test_topic_schema(built_dita: Path, dita: etree.ElementTree, topic_schema: XMLSchema) -> None:
    topic_schema.validate(dita)

```

**Comment by brunato** [comment]
Very good for your case resolution!

Anyway I will add another option to `XMLSchema` and `XMLResource` classes for handling similar cases, when a URN/URL schema location have to be redirected to another URL. In many cases this could be an alternative to export.

The new option will be a dictionary or a function and will be handled by `XMLResource` instances.

**Comment by brunato** [comment]
Hi,
if you will still need to use a custom URI (URN/URL) resolver take a look at the new option **_uri_mapper_**, available within the new major release (v3.0.0).
